server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java

```diff
@@ -1186,48 +1187,89 @@ public IndexMetaData build() {
         }
 
         public static void toXContent(IndexMetaData indexMetaData, XContentBuilder builder, ToXContent.Params params) throws IOException {
+            MetaData.XContentContext context = MetaData.XContentContext.valueOf(params.param(CONTEXT_MODE_PARAM, "API"));
+
             builder.startObject(indexMetaData.getIndex().getName());
 
             builder.field(KEY_VERSION, indexMetaData.getVersion());
             builder.field(KEY_MAPPING_VERSION, indexMetaData.getMappingVersion());
             builder.field(KEY_SETTINGS_VERSION, indexMetaData.getSettingsVersion());
             builder.field(KEY_ALIASES_VERSION, indexMetaData.getAliasesVersion());
             builder.field(KEY_ROUTING_NUM_SHARDS, indexMetaData.getRoutingNumShards());
+
             builder.field(KEY_STATE, indexMetaData.getState().toString().toLowerCase(Locale.ENGLISH));
 
             boolean binary = params.paramAsBoolean("binary", false);
 
             builder.startObject(KEY_SETTINGS);
-            indexMetaData.getSettings().toXContent(builder, new MapParams(Collections.singletonMap("flat_settings", "true")));
+            if (context != MetaData.XContentContext.API) {
+                indexMetaData.getSettings().toXContent(builder, new MapParams(Collections.singletonMap("flat_settings", "true")));
+            }
+            else {
+                indexMetaData.getSettings()
+                    .toXContent(builder, new MapParams(Collections.singletonMap("flat_settings", params.param("flat_settings"))));
```

Comment:
As before the rest of `params` was not passed along to `IndexMetaData.toXContent`, I suspect this may change the serialization and thus introduce a breaking change. I can try and submit another commit with that to see what happen?

Reply:
Actually the original `ClusterState#toXContent` uses the whole `params` https://github.com/elastic/elasticsearch/blob/2e617c3b1026bb3fc4063feeb314c1925e83300d/server/src/main/java/org/elasticsearch/cluster/ClusterState.java#L444. I've updated it.